app/assets/javascripts/icecomm.js

```diff
@@ -1,207 +1,770 @@
+    ic.sendIceCandidate(callerID, event.candidate);
   };
 
-  window.reattachMediaStream = function(to, from) {
-    to.src = from.src;
+  localPeerConnections[callerID].channel.onaddstream = function gotRemoteStream(event){
+    s.addRemoteStream(callerID, event.stream);
   };
-} else {
-  console.log('Browser does not appear to be WebRTC-capable');
+
+  localPeerConnections[callerID].channel.onnegotiationneeded = utils.negotiationNeededHandler;
+  localPeerConnections[callerID].channel.ondatachannel = function(event) {
+    dc.setChannel(callerID, event.channel);
+    dc.addListenersToDataChannel(callerID);
+  }
+}
+
+function createOffer(callerID) {
+  localPeerConnections[callerID].channel.createOffer(function(description) {
+      description.sdp = utils.parseSDP(description.sdp);
+      createDescription(description, callerID, 'offer');
+    }, utils.errorHandler);
 }
 
+function createDescription(description, callerID, offerOrAnswer) {
+  var descriptionObj = {
+    description: description,
+    from: d.getMyID(),
+    // roomSettings: roomSettings,
+    to: callerID
+  };
+
+  if (offerOrAnswer === 'offer') {
+    descriptionObj.serverInfo = serverInfo;
+  }
+  localPeerConnections[callerID].channel.setLocalDescription(description, function() {
+    utils.printDebugMessage('sending ' + offerOrAnswer + ' to ' + callerID);
+    socket.emit(offerOrAnswer, descriptionObj);
+  }, utils.errorHandler);
+}
+
+function removePeer(callerID) {
+  s.removeRemoteStream(callerID);
+  dc.removeDataChannel(callerID);
+  if (localPeerConnections[callerID]) {
+    localPeerConnections[callerID].channel.close();
+    localPeerConnections[callerID] = undefined;
+  }
+}
+
+function addIceCandidate(callerID, candidate) {
+  localPeerConnections[callerID].channel.addIceCandidate(candidate, function() {
+    utils.printDebugMessage('success on setting candidate');
+  }, utils.errorHandler);
+}
+
+function onLeftHandler(callerID) {
+  removePeer(callerID);
+  var options = {};
+  options.callerID = callerID;
+  EventEmitter.trigger('disconnect', options);
+}
+
+function offerEvent(callerID, remoteDescription) {
+  createPeerConnection(callerID);
+
+  if (r.wantsStream()) {
+    addLocalStream(callerID);
+  }
+
+  var offer = new RTCSessionDescription(remoteDescription);
+  setRemoteDescription(callerID, offer, function() {
+    createAnswer(callerID);
+  });
+}
+
+function answerEvent(callerID, remoteDescription) {
+  utils.printDebugMessage('receive answer from ' + callerID);
+  var answer = new RTCSessionDescription(remoteDescription);
+  setRemoteDescription(callerID, answer, function() {
+    utils.printDebugMessage('answer successful');
+  });
+}
+
+function send(callerID, data) {
+  data = JSON.stringify(data);
+  if (callerID === undefined) {
+    for (var callerID in localPeerConnections) {
+      if (dc.isOpen(callerID)) {
+        dc.send(callerID, data);
+      } else {
+        if(dc.isPending(callerID)){
```

Comment:
Added if(dc.isPending(callerID)) check before calling setPending. The isPending function returns false if there is not a dataChannel defined for the callerID - this mirrors the isOpen(callerID) functionality. Previously we were not checking to see if the callerID was valid - we assumed it was and called setPending(callerID,data), which sets dataChannels[callerID].status == "pending". Since dataChannels[callerID] is set to undefined when a user leaves a room, we were trying to set the status attribute of an undefined object, hence the "cannot set property status of undefined" error when trying to send chat messages to all the callerIDs in the localPeerConnections object.